mgr/volumes: clone from snapshot

vshankar opened the pull request:
draft pull request since I've deliberately not included the commit that does a bulk copy of data from a snapshot to the target cloned subvolume -- that code is pretty lame atm.

the rest of the pr is reviewable.

commands/flow:
```
$ ceph fs subvolume create a sub0
$ ceph fs subvolume snapshot create a sub0 snap0 
$ ceph fs subvolume snapshot protect a sub0 snap0
$ ceph fs subvolume snapshot clone a sub0 snap0 clone0
$ ceph fs clone status a clone0
{
  "status": {
    "state": "complete"
  }
}
$ ceph fs subvolume snapshot unprotect a sub0 snap0
```

vshankar commented:
depends on pr #31763 (carries commits)

joscollin commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1105,3 +1109,110 @@ def test_mgr_eviction(self):
+            self._fs_cmd("subvolume", "snapshot", "rm", self.volname, subvolume, snapshot)
+        except CommandFailedError as ce:
+            if ce.exitstatus != errno.EINVAL:
+                raise
```
If success,
```
else:
            raise RuntimeError()
```

joscollin commented on src/pybind/mgr/volumes/fs/fs_util.py:
```diff
@@ -0,0 +1,102 @@
+    # create the given pool
+    command = {'prefix': 'osd pool create', 'pool': pool_name}
+    r, outb, outs = mgr.mon_command(command)
+    if r != 0:
```
drop?

vshankar replied:
sure -- this commit is a part of pr #31763 -- I'll fix it there

ajarr commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -0,0 +1,210 @@
+            pass
+
+    def cancel_job(self, volname):
+        self.lock.acquire()
```
Can this just be,
```
with self.lock():
       self._cancel_job(volume)
```

vshankar replied:
sure

ajarr commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -0,0 +1,210 @@
+        self.lock.release()
+
+    def cancel_all_jobs(self):
+        self.lock.acquire()
```
Can this be,
```
with self.lock():
       for volname in list(self.q):
             self._cancel_job(volname)
```

vshankar replied:
sure

ajarr commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -0,0 +1,210 @@
+    def _cancel_job(self, volname):
+        log.info("cancelling jobs for volume '{0}'".format(volname))
+        try:
+            if not volname in self.q and not volname in self.jobs:
```
in the earlier implementation you returned without checking whether `volname` was in `self.jobs`. Was that a bug?

vshankar replied:
That check already existed in the form of:
```
            if not self.jobs.get(volname, []):		
￼                 return
```

ajarr commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -0,0 +1,210 @@
+        except (KeyError, ValueError):
+            pass
+
+    def cancel_job(self, volname):
```
The naming confuses me. This cancels all jobs of a volume within a lock. Maybe renaming it as `cancel_jobs_with_lock` is more intuitive, or a docstring will help?

vshankar replied:
why would the caller worry if the lock is taken or not? just as the caller does not care if the lock is taken when queuing a job. what's the confusion here? if it's the naming, then `cancel_job_for_volume` is more intuitive. but then, that means all other public APIs need to be suffied with `_for_volume`.

ajarr replied:
Agree, _with_lock suggestion isn't helpful. I think `cancel_jobs` is sufficient.

ajarr commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -0,0 +1,210 @@
+                self.jobs[volname] = []
+            self.cv.notifyAll()
+
+    def _cancel_job(self, volname):
```
The naming confuses me. This cancels all jobs of a volume. Can it be renamed `_cancel_jobs` or something more intuitive.

vshankar replied:
this is not public-facing -- note that leading `_`.

ajarr replied:
I don't follow. Does it matter if the method is public facing or not for it to have a intuitive name? If you think _cancel_job is intuitive, then OK.

vshankar replied:
my comment was related to your suggestion for renaming this to `_cancel_job_with_lock`

this would be called `_cancel_jobs`

vshankar commented on src/pybind/mgr/volumes/fs/operations/subvolume.py:
```diff
@@ -46,7 +46,7 @@ def remove_subvol(fs, vol_spec, group, subvolname):
 
 @contextmanager
-def open_subvol(fs, vol_spec, group, subvolname):
+def open_subvol(fs, vol_spec, group, subvolname, need_complete=True, expected_types=[]):
```
done

vshankar commented on src/pybind/mgr/volumes/fs/operations/clone_index.py:
```diff
@@ -45,6 +47,26 @@ def untrack(self, tracking_id):
         except cephfs.Error as e:
             raise IndexException(-e.args[0], e.args[1])
 
+    def get_clone_entry(self, exclude=[]):
```
sure

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -0,0 +1,210 @@
+        # lock, cv for kickstarting jobs
+        self.lock = threading.Lock()
+        self.cv = threading.Condition(self.lock)
+        # cv for task cancelation
```
Recall that the previous cancel implementation was not thread safe. This change fixes that by removing the extra lock and having a single lock for different condition variables.

ajarr replied:
OK. Sharing the lock between the start and cancel condition objects has made this thread safe.

Can you please mention/explain this change in the commit message?

vshankar replied:
sure

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -0,0 +1,210 @@
+        log.debug("registering async task {0}.{1} with thread {2}".format(volname, job, thread_id))
+        self.jobs[volname].append((job, thread_id))
+
+    def unregister_async_task(self, volname, job, thread_id):
```
task generally refers to an executing entity -- the public APIs are prefixed by `_job` while internally they are referred to as tasks. Guess we should stick using `_job` everywhere? Would that help?

vshankar replied:
and since this class would be a base class for async implementations, some private APIs are named as `_async_task`.

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -0,0 +1,210 @@
+    def reset_cancel(self):
+        self.cancel_event.clear()
+
+class AsyncJob(object):
```
right -- will change to `AsyncJobs`

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -0,0 +1,210 @@
+    Job cancelation is for a job as a whole, i.e., all executing jobs
+    for a job are canceled. Cancelation is poll based -- jobs need to
+    periodically check if cancelation is requested, after which the
+    task should return as soon as possible. Cancelation check is provided
```
ok, so the naming seems to be causing confusion. as mentioned above, a task in just an internal name for a queued (and executed later) job. if this makes sense, then I can improve code comments to make it clear.

ajarr replied:
Yes, code comments should help. Looks like it's not always clear what a job or a task is, https://stackoverflow.com/questions/3073948/job-task-and-process-whats-the-difference?answertab=votes#tab-top

vshankar replied:
well, the link you gave has a very generic explanation. However, I do agree that job vs task can be more clearly defined in the async module. I'll get that sorted out.

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -0,0 +1,210 @@
+            self.jobs.pop(vol)
+        return next_job
+
+    def register_async_task(self, volname, job, thread_id):
```
ditto -- task (internal) vs job (external/user).

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py:
```diff
@@ -0,0 +1,139 @@
+        try:
+            fd = self.fs.open(self.config_path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, self.mode)
+            wrote = 0
+            while True:
```
`conf_data.read()` might not return the whole buffered string -- I observed this with py2.

ajarr commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -0,0 +1,210 @@
+
+    def get_next_job(self, volname, running_jobs):
+        """
+        get the next (job, task) tuple for asynchronous execution. called under `self.lock`.
```
Why is this method returning a tuple? Isn't returning just the job/task sufficient? Both `get_next_clone_entry()` and `get_trash_entry_for_volume()` return (0, job/task).

vshankar replied:
the return value is required since there may be no more jobs (entries) to process.

vshankar commented:
@ajarr thanks for the review. I pushed an update with some of your comments addressed (the job vs task thing has _not_ been addressed, I'll work on that). However, I added the bulk copy feature (which I left out in my original commit).

So, this can be tried out.

vshankar commented:
@ajarr rebased over pr #31763. Fixed minor issues with clone operation getting interrupted.

vshankar commented:
rebase and updated. please take a look.

vshankar commented:
@ajarr pinging for reviews?

(I'm updating doc/cephfs/fs-volumes.rst w/ clone specific commands)

vshankar commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at http://docs.ceph.com/ceph-prs/32030/

vshankar commented:
minor update to doc.

vshankar commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at http://docs.ceph.com/ceph-prs/32030/

ajarr commented on doc/cephfs/fs-volumes.rst:
```diff
@@ -172,5 +172,86 @@ List snapshots of a subvolume using::
+
+  $ ceph fs subvolume snapshot unprotect <vol_name> <subvol_name> <snap_name> [--group_name <subvol_group_name>]
+
+Note that unprotecting a snapshot would fail if there are pending or in progress clone operations. Also note that,
```
Can you unprotect a snapshot whose clone is in 'failed' state? Else if a clone fails, then you can't ever remove the snapshot and the original subvolume.

vshankar replied:
sure thing! on failure, the (incomplete) clone is "detached" from the source (snapshot).

ajarr commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -96,3 +231,70 @@ def list_snapshots(self):
+            raise VolumeException(-errno.ENOENT, "snapshot '{0}' does not exist".format(snapname))
+        if not self.is_snapshot_protected(snapname):
+            raise VolumeException(-errno.EEXIST, "snapshot '{0}' is not protected".format(snapname))
+        if self.has_pending_clones(snapname):
```
can you unprotect snapshot if the clone has failed?

vshankar replied:
ditto

vshankar commented on src/pybind/mgr/volumes/module.py:
```diff
@@ -163,6 +163,46 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
+                   'name=sub_name,type=CephString '
+                   'name=snap_name,type=CephString '
+                   'name=target_sub_name,type=CephString '
+                   'name=pool_layout,type=CephString,req=false '
```
isn't it obvious that `pool_layout` is for the target subvolume (clone)?

ajarr replied:
OK

vshankar commented on src/pybind/mgr/volumes/fs/volume.py:
```diff
@@ -270,6 +271,107 @@ def list_subvolume_snapshots(self, **kwargs):
+                with open_subvol(fs_handle, self.volspec, target_group, target_subvolname, need_complete=False):
+                    raise VolumeException(-errno.EEXIST, "subvolume '{0}' exists".format(target_subvolname))
+            except VolumeException as ve:
+                if ve.errno == -errno.ENOTSUP:
```
ENOTSUP is raised in `src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py::open()`.

the log message probably could have been more meaningful. This happens when cloning to an existing subvolume (normal subvolume or a clone).

vshankar replied:
guess this was not required -- I had initially handled a case where cloning to an existing clone did not return an error. this was leftover from that (ENOTSUP gets returned when a subvolume for a given type ("subvolume" or "clone") exists.

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1262,3 +1318,636 @@ def test_subvolume_rm_with_snapshots(self):
+        # remove snapshot should fail since the snapshot is protected
+        try:
+            self._fs_cmd("subvolume", "snapshot", "rm", self.volname, subvolume, snapshot)
+        except CommandFailedError as ce:
```
yeh -- had this change locally. thanks!

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1262,3 +1318,636 @@ def test_subvolume_rm_with_snapshots(self):
+
+        # remove subvolumes
+        self._fs_cmd("subvolume", "rm", self.volname, subvolume)
+        self._fs_cmd("subvolume", "rm", self.volname, clone)
```
removed when filesystem gets removed?

vshankar commented:
@ajarr fixed and updated. please take a look.

I added a commit (dd2c50d13d2f41af8a6086927bc1e9360dd14339) to allow removing a failed clone and added a test case for that (plus, doc updated).

vshankar commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at http://docs.ceph.com/ceph-prs/32030/

ajarr commented:
Thanks, Venky! Code looks great.
It'll be nice to mention more info about the clone in the doc.
* Cloning only copies over regular files, directories and symbolic links as symbolic links
* It preserves ownership, mode, but not the xattrs and timestamps (?)
* If the clone fails, you'll have to delete the partial clone and retry afresh.

vshankar commented:
> Thanks, Venky! Code looks great.
> It'll be nice to mention more info about the clone in the doc.
> 
> * Cloning only copies over regular files, directories and symbolic links as symbolic links
> * It preserves ownership, mode, but not the xattrs and timestamps (?)

`mode is preserved`, but not uid/gid for each inode (although for the subvolume mode, uid and gid are synced from the source subvolume). I think clone should preserver uid/gid for each inode.

> * If the clone fails, you'll have to delete the partial clone and retry afresh.

vshankar commented:
> `mode` is preserved, but not uid/gid for each inode (although for the subvolume mode, uid and gid are synced from the source subvolume). I think clone should preserver uid/gid for each inode.
> 
> >

BTW, the reason for not sycning uid/gid was that only the owner/superuser would have the perms to change file ownership. Since fs handles in mgr/volumes have mount uid/gid set to 0, mgr/volumes should be able to chown() at will...

vshankar commented:
jenkins render docs

vshankar commented:
@ajarr -- couple of changes

* added ownership sync support for directories and regular files. For symlinks, there's no `lchown()` python binding.
* updated doc mentioning the above limitation along with the limitation that clone does not sync inode timestamps. also, mentioned about deleting clone in case of failure.

The next step would be to add `lchown` and `[l]utimes` support in cephfs python bindings.

ceph-jenkins commented:
Doc render available at http://docs.ceph.com/ceph-prs/32030/

joscollin commented:
Looks like the meaning of `--force` is changed when a snapshot it is protected?

```
$ ./bin/ceph fs subvolume snapshot rm myfs mysubvol snap1 --group_name mygroup
Error EINVAL: snapshot 'snap1' is protected
$ ./bin/ceph fs subvolume snapshot rm myfs mysubvol snap1 --group_name mygroup --force
Error EINVAL: snapshot 'snap1' is protected
```
That's not how `--force` worked before.

vshankar commented:
> Looks like the meaning of `--force` is changed when a snapshot it is protected?
> 
> ```
> $ ./bin/ceph fs subvolume snapshot rm myfs mysubvol snap1 --group_name mygroup
> Error EINVAL: snapshot 'snap1' is protected
> $ ./bin/ceph fs subvolume snapshot rm myfs mysubvol snap1 --group_name mygroup --force
> Error EINVAL: snapshot 'snap1' is protected
> ```
> 
> That's not how `--force` worked before.

... and it should not work when snapshot is protected. That's foot shooting.

ajarr commented:
All the tests in test_volumes passed for me using teuthology locally.  Now, running the test_volumes test suite using teuthology in the upstream CI.

ajarr commented:
http://pulpito.front.sepia.ceph.com/rraja-2020-01-30_17:03:37-fs-wip-rraja-testing-32030-distro-basic-smithi/

vshankar commented:
> http://pulpito.front.sepia.ceph.com/rraja-2020-01-30_17:03:37-fs-wip-rraja-testing-32030-distro-basic-smithi/

one failure is due to cluster warn message when the pool is full -- one of the test fills up a pool to fail a clone operation. for this, the log message needs to be whitelisted.

for the other failure, the clone completed fast enough before the test could unprotect the snapshot and expect the unprotect operation to fail. we should probably increase the data set size.

I'll try and post the results.

vshankar commented:
teuthology run with test case fixes: http://pulpito.ceph.com/vshankar-2020-01-31_08:29:53-fs-wip-mgr-volumes-clone-distro-basic-smithi/

vshankar commented:
@ajarr fixed and updated. I haven't put the test case race as a comment. I guess it's ok -- we can always do that later...

ajarr commented:
> @ajarr fixed and updated. I haven't put the test case race as a comment. I guess it's ok -- we can always do that later...

Sounds good.

batrick commented:
@ajarr please make sure your merge commit message includes the `Reviewed-by: ....` qualifiers.

batrick commented:
(for future merges)

ajarr commented:
> @ajarr please make sure your merge commit message includes the `Reviewed-by: ....` qualifiers.

Oops! I missed that. Thanks! I will add it next time.
